### ISSUE TITLE
Type narrowing regression

### ISSUE THREAD

**Issue by mariusmuja** [microsoft/pyright#2159]
**Describe the bug**
When type-checking the following code:

```python
from typing import Tuple, Union

MsgA = Tuple[int, str]
MsgB = Tuple[int, float]

Msg = Union[MsgA, MsgB]


def bar(b: MsgB):
    pass

def foo(a: Msg):
    assert type(a) is MsgB
    bar(a)
```

the following error is displayed:
```
test.py:14:9 - error: Argument of type "Msg" cannot be assigned to parameter "b" of type "MsgB" in function "bar"
    Type "Msg" cannot be assigned to type "MsgB"
      Tuple entry 2 is incorrect type
        "str" is incompatible with "float" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
```

**Expected behavior**
The code above type-checks ok in pyright 1.1.158, and fails in 1.1.159 and 1.1.160.

**Comment by erictraut** [comment]
I don't understand what this code is trying to do. Both `MsgA` and `MsgB` are type aliases that refer to specialized versions of the generic tuple class. They are not classes that can be instantiated. That means `type(a)` will never return `MsgB`, nor will it return `tuple[int, float]`.

```python
a: MsgA = (0, "")
print(type(a)) # <class 'tuple'>

b: MsgB = (0, 0.0)
print(type(b)) # <class 'tuple'>
```

I think pyright is correct in not applying any type narrowing here, since the code doesn't make sense.

Mypy also doesn't apply type narrowing for this sample and emits the same error as pyright.

**Comment by mariusmuja** [comment]
He's the example without the type aliases:

```python
from typing import Tuple, Union

def bar(b: float):
    pass

def foo(a: Union[Tuple[str], Tuple[float]]):
    assert type(a) is Tuple[float]
    reveal_type(a)
    bar(a[0])
```

pyright outputs:
```
test.py:9:17 - info: Type of "a" is "Tuple[str] | Tuple[float]"
test.py:10:9 - error: Argument of type "str | float" cannot be assigned to parameter "b" of type "float" in function "bar"
    Type "str | float" cannot be assigned to type "float"
      "str" is incompatible with "float" (reportGeneralTypeIssues)
```

I expected pyright to narrow the type of `a` to just `Tuple[float]` after the `assert type(a) is Tuple[float]` line, as it's been doing until version 1.1.158.

**Comment by jakebailey** [comment]
`assert type(a) is Tuple[float]` won't work at runtime; that's not the right way to verify that the tuple contains a float (or has any given shape, actually).

For example, try running this code:

```
from typing import Tuple, Union

def foo(a: Union[Tuple[str], Tuple[float]]):
    assert type(a) is Tuple[str]
    print(a)
    
foo(("I'm no float",))
```

This fails the assertion, even though based on your original example it would seem that it should. See: https://glot.io/snippets/g16h9ln0xf

If you want to narrow the type of a tuple, you're going to have to either write a type guard to distinguish them, or maybe check the type of one of the items, like:

```py
from typing import Tuple, Union

def foo(a: Union[Tuple[str], Tuple[float]]):
    assert isinstance(a[0], float)
    reveal_type(a[0])  # float
```

But Python's not going to do anything special to verify `a is Tuple[int, float]` or any other typing property like that, as far as I'm aware. (I believe the assertion fails because `typing.Tuple` is not `tuple` at runtime like that. At most, maybe in more recent versions, they are the same, in which case the type check would _always_ work?)

If this worked previously, that must have been a bug...

**Comment by erictraut** [comment]
Jake is correct. At runtime, type arguments for generic types are effectively ignored, so `type(a)` will always return `tuple`.

If this previously worked without generating a type error, then it was a bug. It is working correctly now.

You can use `TypeGuard` (defined in PEP 647) to handle this case.

```python
from typing import Any, Tuple, Type, TypeGuard, TypeVar, Union

_T = TypeVar("_T")

def is_tuple_of(x: Tuple[Any], typ: Type[_T]) -> TypeGuard[Tuple[_T]]:
    return x[0] is type

def foo(a: Union[Tuple[str], Tuple[float]]):
    assert is_tuple_of(a, float)
    reveal_type(a)  # Tuple[float]
    bar(a[0])
```

**Comment by mariusmuja** [comment]
> assert type(a) is Tuple[float] won't work at runtime; that's not the right way to verify that the tuple contains a float (or has any given shape, actually).

I wasn't interested in this working at runtime, only at type-checking time, the example I gave is a minimal example to reproduce the issue. My use case us more similar to the first example, a tagged union of tuples where the first element of each tuple is the tag:

```python
from typing import Tuple, Union, TYPE_CHECKING, cast

TagA = 1
TagB = 2

MsgA = Tuple[int, str]
MsgB = Tuple[int, float]

Msg = Union[MsgA, MsgB]


def handle_a(a: MsgA):
    pass

def handle_b(b: MsgB):
    pass

def handle(m: Msg):
    if m[0] == TagA:
        if TYPE_CHECKING: assert type(m) is MsgA
        handle_a(m)
    elif m[0] == TagB:
        if TYPE_CHECKING: assert type(m) is MsgA
        handle_b(m)
```

Indeed, using `TypeGuard` is the nicer way of doing this, however that requires a function call and function calls as expensive in python (at least in my use case I'd want to avoid a extra function call, which is why I'm using the ` if TYPE_CHECKING:` hack).

Turns out I can change it to:
```python
def handle(m: Msg):
    if m[0] == TagA:
        if TYPE_CHECKING: m = cast(MsgA, m)
        handle_a(m)
    elif m[0] == TagB:
        if TYPE_CHECKING: m = cast(MsgB, m)
        handle_b(m)
```

Ideally I would be able to write just:
```python
MsgA = Tuple[Literal[1], str]
MsgB = Tuple[Literal[2], float]

def handle(m: Msg):
    if m[0] == 1:
        handle_a(m)
    elif m[0] == 2:
        handle_b(m)
```
and pyright would narrow the type on `m` to `MsgA` and `MsgB` in the respective branches.

**Comment by erictraut** [comment]
Type checkers cannot make up new semantics for existing constructs in Python, at least not for statements that execute at runtime. It would largely defeat the purpose of type checking if semantics and behaviors differed between runtime and type analysis type.

Of course, you're free to do whatever you want in your own code, but I would recommend against using `TYPE_CHECKING` conditionals or `cast` calls if you can help it. These both cause the runtime behavior and type checking behaviors to diverge, so they make your code more fragile.

**Comment by mariusmuja** [comment]
Turns out the following type checks successfully with `mypy` (https://mypy.readthedocs.io/en/stable/literal_types.html#tagged-unions):

```python
from typing import Tuple, Union, Literal

MsgA = Tuple[Literal[1], str]
MsgB = Tuple[Literal[2], float]

Msg = Union[MsgA, MsgB]

def handle_a(a: MsgA):
    pass

def handle_b(b: MsgB):
    pass


def handle(m: Msg):
    if m[0] == 1:
        handle_a(m)
    elif m[0] == 2:
        handle_b(m)
```

Could this be supported by pyright as well? Should I open a different issue for this?

**Comment by erictraut** [comment]
Yes, please open a separate enhancement request. This will require support for a new type narrowing pattern. Since mypy already supports this pattern, pyright should probably support it as well.
